### PR TITLE
add compilation restrictions

### PR DESCRIPTION
**Pull request by gretzke**

### CODE REVIEW

File: src/interfaces/IPoolInitializer.sol [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+
+interface IPoolInitializer {
+    /// @notice Initialize a Uniswap v4 Pool
+    /// @dev If the pool is already initialized, this function will not revert and just return type(int24).max
+    /// @param key the PoolKey of the pool to initialize
+    /// @param sqrtPriceX96 the initial sqrtPriceX96 of the pool
```
Comment (louisbrown0212) [inline src/interfaces/IPoolInitializer.sol:10]: nit: It can be capitalized for better readability
```suggestion
    /// @param key The PoolKey of the pool to initialize
    /// @param sqrtPriceX96 The initial sqrtPriceX96 of the pool
```

File: src/PositionDescriptor.sol [hunk 2]
```diff
@@ -87,18 +85,12 @@ contract PositionDescriptor is IPositionDescriptor {
         );
     }
 
-    /// @notice Returns true if currency0 has higher priority than currency1
-    /// @param currency0 The first currency address
-    /// @param currency1 The second currency address
-    /// @return flipRatio True if currency0 has higher priority than currency1
-    function flipRatio(address currency0, address currency1) public view returns (bool) {
+    /// @inheritdoc IPositionDescriptor
+    function flipRatio(address currency0, address currency1) public view override returns (bool) {
```
Comment (louisbrown0212) [inline src/PositionDescriptor.sol:89]: nit: `override` is not necessary.

File: src/interfaces/IPositionManager.sol [hunk 3]
```diff
@@ -40,7 +51,11 @@ interface IPositionManager is INotifier, IImmutableState {
     function getPositionLiquidity(uint256 tokenId) external view returns (uint128 liquidity);
 
     /// @param tokenId the ERC721 tokenId
-    /// @return PositionInfo a uint256 packed value holding information about the position including the range (tickLower, tickUpper)
```
Comment (dianakocsis) [inline src/interfaces/IPositionManager.sol:43]: needs @notice tag along with the other 2

**Comment by dianakocsis** [comment]
could be removed
https://github.com/Uniswap/v4-periphery/blob/d1d087f0794116a851d44a1b62f998c27cb39c8a/test/shared/Planner.sol#L6

**Comment by dianakocsis** [comment]
getting warnings on these:
https://github.com/Uniswap/v4-periphery/blob/d1d087f0794116a851d44a1b62f998c27cb39c8a/test/mocks/MockBadSubscribers.sol#L51
https://github.com/Uniswap/v4-periphery/blob/d1d087f0794116a851d44a1b62f998c27cb39c8a/test/mocks/MockBadSubscribers.sol#L95
https://github.com/Uniswap/v4-periphery/blob/d1d087f0794116a851d44a1b62f998c27cb39c8a/test/mocks/MockSubscriber.sol#L50

File: src/interfaces/IPoolInitializer.sol [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+pragma solidity ^0.8.0;
+
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+
+interface IPoolInitializer {
+    /// @notice Initialize a Uniswap v4 Pool
+    /// @dev If the pool is already initialized, this function will not revert and just return type(int24).max
+    /// @param key The PoolKey of the pool to initialize
+    /// @param sqrtPriceX96 The initial sqrtPriceX96 of the pool
+    /// @return tick The current tick of the pool
```
Comment (hensha256) [inline src/interfaces/IPoolInitializer.sol:11]:
```suggestion
    /// @return tick The current tick of the pool, or type(int24).max if the pool creation failed, or the pool already existed
```

File: test/mocks/MockReenterHook.sol [hunk 5]
```diff
@@ -21,8 +22,8 @@ contract MockReenterHook is BaseTestHooks {
         }
         (bytes4 selector, address owner, uint256 tokenId) = abi.decode(functionSelector, (bytes4, address, uint256));
 
-        if (selector == posm.transferFrom.selector) {
-            posm.transferFrom(owner, address(this), tokenId);
+        if (selector == IERC721(address(posm)).transferFrom.selector) {
```
Comment (hensha256) [inline test/mocks/MockReenterHook.sol:25]: why doesnt IPositionManager just inherit from IERC721?

File: test/position-managers/NativeToken.t.sol [hunk 6]
```diff
@@ -20,12 +20,11 @@ import {SafeCast} from "@uniswap/v4-core/src/libraries/SafeCast.sol";
 import {Position} from "@uniswap/v4-core/src/libraries/Position.sol";
 
 import {IERC20} from "forge-std/interfaces/IERC20.sol";
+import {IERC721} from "forge-std/interfaces/IERC721.sol";
 import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
```
Comment (hensha256) [inline test/position-managers/NativeToken.t.sol:24]: why the change from OZ to forge?

File: test/position-managers/Permit.t.sol [hunk 7]
```diff
@@ -226,7 +226,7 @@ contract PermitTest is Test, PosmTestSetup {
 
         // alice revokes the nonce
         vm.prank(alice);
-        lpm.revokeNonce(nonce);
+        UnorderedNonce(address(lpm)).revokeNonce(nonce);
```
Comment (hensha256) [inline test/position-managers/Permit.t.sol:229]: should there be an IUnorderedNonce that IPositionManager inherits from?

File: test/position-managers/PositionManager.multicall.t.sol [hunk 8]
```diff
@@ -422,8 +421,8 @@ contract PositionManagerMulticallTest is Test, Permit2SignatureHelpers, PosmTest
 
         // bob front-runs the permits
         vm.startPrank(bob);
-        lpm.permit(charlie, permit0, sig0);
-        lpm.permit(charlie, permit1, sig1);
+        Permit2Forwarder(address(lpm)).permit(charlie, permit0, sig0);
```
Comment (hensha256) [inline test/position-managers/PositionManager.multicall.t.sol:424]: Same here, shouldnt this be in the IPosm interface?

File: test/position-managers/PositionManager.modifyLiquidities.t.sol [hunk 9]
```diff
@@ -837,6 +837,16 @@ contract PositionManagerModifyLiquiditiesTest is Test, PosmTestSetup, LiquidityF
         assertEq(lpm.getPositionLiquidity(tokenId), initialLiquidity + newLiquidity);
     }
 
+    struct BalanceDiff {
+        uint256 before;
+        uint256 _after;
```
Comment (hensha256) [inline test/position-managers/PositionManager.modifyLiquidities.t.sol:842]: is the underscore needed? would prefer `.after` and `.0` and `.1`

File: foundry.toml [hunk 10]
```diff
@@ -1,15 +1,25 @@
 [profile.default]
 out = 'foundry-out'
 solc_version = '0.8.26'
-optimizer_runs = 1
+optimizer_runs = 30000
```
Comment (hensha256) [inline foundry.toml:4]: can we do 4444444 for things that arent position manager?

File: test/mocks/MockReenterHook.sol (hunk 5 above)
Comment (gretzke) [reply test/mocks/MockReenterHook.sol:25]: I can't really use an interface that defines the events because the solmate ERC721 implementation already includes the events and doesn't have an interface. So if `IPositionManager` inherits from `IERC721` I get an error that the event with the same name and parameter types was defined twice. The only way around that would be to create a custom ERC721 interface without the events, so I opted for casting in the tests instead

File: test/position-managers/NativeToken.t.sol (hunk 6 above)
Comment (gretzke) [reply test/position-managers/NativeToken.t.sol:24]: Every other test file was using forge for the token interfaces so I wanted to be consistent

File: test/position-managers/Permit.t.sol (hunk 7 above)
Comment (gretzke) [reply test/position-managers/Permit.t.sol]: added the interface

File: test/position-managers/PositionManager.modifyLiquidities.t.sol (hunk 9 above)
Comment (gretzke) [reply test/position-managers/PositionManager.modifyLiquidities.t.sol:842]: after is a reserved keyword, also you cannot use pure numbers as variable names

File: test/shared/Deploy.sol [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        bytes memory salt
+    ) internal returns (IPositionManager manager) {
+        bytes memory args = abi.encode(poolManager, permit2, unsubscribeGasLimit, positionDescriptor_, wrappedNative);
+        bytes memory initcode = abi.encodePacked(vm.getCode("PositionManager.sol:PositionManager"), args);
+        assembly {
+            manager := create2(0, add(initcode, 0x20), mload(initcode), salt)
+        }
+    }
+
+    // no interface available, wrap in contract, this results in both versions being compiled, that's why we are importing the default version explicitly from the json file
```
[the remark this answers is not in the document]
Comment (gretzke) [reply test/shared/Deploy.sol]: outdated now

File: src/interfaces/IPoolInitializer.sol [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+pragma solidity ^0.8.0;
+
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+
+interface IPoolInitializer {
+    /// @notice Initialize a Uniswap v4 Pool
+    /// @dev If the pool is already initialized, this function will not revert and just return type(int24).max
+    /// @param key The PoolKey of the pool to initialize
+    /// @param sqrtPriceX96 The initial sqrtPriceX96 of the pool
+    /// @return tick The current tick of the pool, or type(int24).max if the pool creation failed, or the pool already existed
```
Comment (dianakocsis) [inline src/interfaces/IPoolInitializer.sol:11]: remove `tick` here too!

File: test/position-managers/PositionManager.modifyLiquidities.t.sol (hunk 9 above)
Comment (hensha256) [reply test/position-managers/PositionManager.modifyLiquidities.t.sol:842]: sad

File: src/base/UnorderedNonce.sol [hunk 13]
```diff
@@ -1,11 +1,11 @@
...
+
 /// @title Unordered Nonce
 /// @notice Contract state and methods for using unordered nonces in signatures
-contract UnorderedNonce {
-    error NonceAlreadyUsed();
-
+contract UnorderedNonce is IUnorderedNonce {
     /// @notice mapping of nonces consumed by each address, where a nonce is a single bit on the 256-bit bitmap
     /// @dev word is at most type(uint248).max
     mapping(address owner => mapping(uint256 word => uint256 bitmap)) public nonces;
```
Comment (hensha256) [inline src/base/UnorderedNonce.sol:11]: cant this natspec go in the interface too? mayve inheritdoc gets confused with mappings

File: test/position-managers/PositionManager.modifyLiquidities.t.sol [hunk 14]
```diff
@@ -906,60 +917,68 @@ contract PositionManagerModifyLiquiditiesTest is Test, PosmTestSetup, LiquidityF
...
+            Balance memory expected;
+            {
+                bool currency0IsFOT = fotKey.currency0 == Currency.wrap(address(fotToken));
+                uint256 expectedFee = (currency0IsFOT ? amount0 : amount1).calculatePortion(bips);
+                (expected._0, expected._1) = currency0IsFOT
+                    ? (balance0.before - balance0._after - expectedFee, balance1.before - balance1._after)
+                    : (balance0.before - balance0._after, balance1.before - balance1._after - expectedFee);
+            }
+            assertEq(expected._0, balance0PM._after - balance0PM.before);
+            assertEq(expected._1, balance1PM._after - balance1PM.before);
```
Comment (hensha256) [inline test/position-managers/PositionManager.modifyLiquidities.t.sol:955]: i guess given `_0` `_1` and `_after` have to have `_`, I'd like `before` to be `_before`

File: test/shared/Deploy.sol [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    function stateView(address poolManager) internal returns (StateView stateView_) {
+        bytes memory args = abi.encode(poolManager);
+        bytes memory initcode = abi.encodePacked(vm.getCode("foundry-out/StateView.sol/StateView.default.json"), args);
+        assembly {
+            stateView_ := create(0, add(initcode, 0x20), mload(initcode))
```
Comment (0age) [inline test/shared/Deploy.sol:32]: why aren't these ones using create2 as well?

File: script/DeployStateView.s.sol [hunk 16]
```diff
@@ -5,16 +5,16 @@ import "forge-std/console2.sol";
 import "forge-std/Script.sol";
 
 import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
```
Comment (saucepoint) [inline script/DeployStateView.s.sol:7]: unused import no?

File: script/DeployV4Quoter.s.sol [hunk 17]
```diff
@@ -5,16 +5,16 @@ import "forge-std/console2.sol";
 import "forge-std/Script.sol";
 
 import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
```
Comment (saucepoint) [inline script/DeployV4Quoter.s.sol:7]: unused import now

File: src/interfaces/IPositionDescriptor.sol [hunk 18]
```diff
@@ -2,13 +2,37 @@
...
+    /// @param currency0 The first currency address
+    /// @param currency1 The second currency address
+    /// @return True if currency0 has higher priority than currency1
+    function flipRatio(address currency0, address currency1) external view returns (bool);
+
+    /// @notice Returns the priority of a currency.
+    /// For certain currencies on mainnet, the smaller the currency, the higher the priority
+    /// @param currency The currency address
+    /// @return The priority of the currency
+    function currencyRatioPriority(address currency) external view returns (int256);
```
Comment (saucepoint) [inline src/interfaces/IPositionDescriptor.sol:28]: can we document what a negative number means vs a positive number?

File: src/interfaces/IPoolInitializer.sol [hunk 19]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+
+interface IPoolInitializer {
+    /// @notice Initialize a Uniswap v4 Pool
+    /// @dev If the pool is already initialized, this function will not revert and just return type(int24).max
+    /// @param key The PoolKey of the pool to initialize
+    /// @param sqrtPriceX96 The initial sqrtPriceX96 of the pool
```
Comment (saucepoint) [inline src/interfaces/IPoolInitializer.sol:10]:
```suggestion
    /// @param sqrtPriceX96 The initial starting price of the pool, expressed as a sqrtPriceX96
```

Comment (saucepoint) [reply src/interfaces/IPoolInitializer.sol:10]: for clarity -- makes it nice downstream where we're generating reference documentation from natspec 😄

File: src/interfaces/IStateView.sol [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+
+import {StateLibrary} from "@uniswap/v4-core/src/libraries/StateLibrary.sol";
+import {PoolId} from "@uniswap/v4-core/src/types/PoolId.sol";
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {Position} from "@uniswap/v4-core/src/libraries/Position.sol";
+import {IImmutableState} from "../interfaces/IImmutableState.sol";
+
+/// @title StateView Interface
+/// @notice A view only contract wrapping the StateLibrary.sol library for reading storage in v4-core.
+interface IStateView is IImmutableState {
```
Comment (saucepoint) [inline src/interfaces/IStateView.sol:12]: just a thought, but not married to this suggestion at all
```suggestion
/// @notice A view only contract wrapping the StateLibrary.sol library for reading storage in v4-core.
/// @dev the contract is intended for offchain clients; reading state onchain should use StateLibrary.sol directly
interface IStateView is IImmutableState {
```

File: test/position-managers/PositionManager.modifyLiquidities.t.sol [hunk 21]
```diff
@@ -906,60 +917,68 @@ contract PositionManagerModifyLiquiditiesTest is Test, PosmTestSetup, LiquidityF
...
-        bool positionIsEntirelyInOtherToken = currency0IsFOT
-            ? tickUpper <= TickMath.getTickAtSqrtPrice(sqrtPriceX96)
-            : tickLower > TickMath.getTickAtSqrtPrice(sqrtPriceX96);
-
-        if (bips == 10000 && !positionIsEntirelyInOtherToken) {
+        // needed to remove variables because of stack too deep
+        // bool currency0IsFOT = fotKey.currency0 == Currency.wrap(address(fotToken));
+        // bool positionIsEntirelyInOtherToken = currency0IsFOT
+        //     ? tickUpper <= TickMath.getTickAtSqrtPrice(sqrtPriceX96)
+        //     : tickLower > TickMath.getTickAtSqrtPrice(sqrtPriceX96);
```
Comment (saucepoint) [inline test/position-managers/PositionManager.modifyLiquidities.t.sol:924]: deadcode or documentation?

**Review by saucepoint — COMMENTED** [review]
looks pretty good to me; did NOT look at PositionDescriptor.t.sol btw 😅

File: src/base/UnorderedNonce.sol (hunk 13 above)
Comment (snreynolds) [reply src/base/UnorderedNonce.sol:11]: could, but then i will need to define another function for `nonces`

Comment (snreynolds) [reply src/base/UnorderedNonce.sol:11]: & its a bit odd bc in the interface it will be declared as external but here it is a public variable

Comment (snreynolds) [reply src/base/UnorderedNonce.sol:9]: I made this change so you could see, but lmk if you want me to go back to this. idc either way tbh

File: test/shared/Deploy.sol (hunk 15 above)
Comment (gretzke) [reply test/shared/Deploy.sol:32]: all contracts that were instantiated using `new Contract()` I used `create` for the deployer, if they were instantiated  with a salt though, I used create2

Comment (snreynolds) [reply test/shared/Deploy.sol]: cool, i changed it to use create2 for consistency

File: src/base/UnorderedNonce.sol (hunk 13 above)
Comment (hensha256) [reply src/base/UnorderedNonce.sol:9]: yeah i like it! its normal that public functions have to be external in interfaces - just good for people to be able to use the interface to access all public information

Comment (hensha256) [reply src/base/UnorderedNonce.sol:9]: just need to add ` /// @inheritdoc IUnorderedNonce` here now

File: src/interfaces/IPermit2Forwarder.sol [hunk 22]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IAllowanceTransfer} from "permit2/src/interfaces/IAllowanceTransfer.sol";
+
+/// @notice Permit2Forwarder allows permitting this contract as a spender on permit2
```
Comment (dianakocsis) [inline src/interfaces/IPermit2Forwarder.sol:6]: this is the exact same natspec as Permit2Forwarder. I feel like the top of interfaces should all be the same, maybe like how it is for IPositionManager.sol. 
`/// @title IPermit2Forwarder
 /// @notice Interface for the Permit2Forwarder contract`
 and then we can follow this standard across all of our interfaces, and then the natspec at the top of the contracts could go in more detail?

**Review by dianakocsis — APPROVED** [review]
